File: js/chart/zones.js
/* zones.js — horizontal support/resistance BANDS, not lines.
 *
 * A port of sim/tl/zones.py, compared zone-for-zone in tests/test_zone_parity.py.
 *
 * A line is a claim about one price. A zone is a claim about a region, and it is
 * the more honest object: price does not turn at 1.16847, it turns somewhere in
 * a band a few tenths of an ATR wide — which is how every hand-annotated chart
 * draws it ("Supply", "Demand", "Seller Zone", "Buyer Zone").
 *
 * ROLE IS ASSIGNED FROM PRICE, not from the pivot type. A zone built from swing
 * highs is resistance while price is below it and support once price is above
 * it; hard-coding the role from its pivots would throw that flip away.
 */

import { atrSeries } from './tlengine.js';
import { findPivots } from './trendlines.js';

export const SUPPORT = 'support';
export const RESISTANCE = 'resistance';

export const DEFAULT_ZONE_PARAMS = {
  strengthPivots: 3,
  lookback: 500,
  clusterAtr: 0.35,
  /* 2 rather than 3: a clean double top IS a level a human draws, and score()
     already gives it 11 points against 35 for four touches, so a weak zone
     loses on merit instead of being excluded before it can compete. */
  minTouches: 2,
  minSeparation: 8,
  maxWidthAtr: 1.2,
  maxZones: 6,
  minStrength: 25,
  /* How far price travelled AWAY from a pivot before turning back, in ATR. A
     level that produced 3 ATR bounces is not the same object as one price
     grazed and drifted from, and a raw touch count cannot tell them apart. */
  reactionBars: 20,
  reactionFullAtr: 2.0,
  /* "12 ATR" meant 19% of the actual price range on USDJPY 1h and 54% on
     XAUUSD 4h -- a trader draws what is ON THE CHART, not a count of ATR. The
     allowance is the LARGER of the ATR budget and a fraction of the range price
     actually covered; the ATR term is a floor for unusually quiet windows. */
  maxDistanceAtr: 12,
  maxDistanceRange: 0.75,
};

const round2 = (v) => Math.round(v * 100) / 100;

export class Zone {
  constructor(o) { Object.assign(this, o); }

  get mid() { return 0.5 * (this.low + this.high); }

  contains(price) { return price >= this.low && price <= this.high; }

  /** Resistance below, support above; inside the band the nearer edge decides. */
  roleAt(price) {
    if (price < this.low) return RESISTANCE;
    if (price > this.high) return SUPPORT;
    return (price - this.low) > (this.high - price) ? SUPPORT : RESISTANCE;
  }

  distanceAtr(price, atr) {
    if (!atr) return NaN;
    if (this.contains(price)) return 0;
    const d = price < this.low ? this.low - price : price - this.high;
    return d / atr;
  }
}

/* Median excursion away from a zone's pivots, in ATR. Median, not mean, so one
   violent bounce cannot carry a level that otherwise did nothing.
   CAUSAL: every bar read lies between the pivot and `i`, and the window is
   clipped at `i` so a recent pivot is not credited with an excursion that has
   not happened yet. */
function reactionAtr(kept, bars, atr, i, p) {
  const out = [];
  // JS `kept` holds {price, i, kind} objects; the Python mirror uses tuples
  for (const lv of kept) {
    const price = lv.price, k = lv.i, kind = lv.kind;
    const a = k < atr.length ? atr[k] : NaN;
    if (!(Number.isFinite(a) && a > 0)) continue;
    const end = Math.min(i, k + p.reactionBars);
    if (end <= k) continue;
    let move;
    if (kind === 'low') {
      let hi = -Infinity;
      for (let m = k + 1; m <= end; m++) if (bars[m].h > hi) hi = bars[m].h;
      move = hi - price;
    } else {
      let lo = Infinity;
      for (let m = k + 1; m <= end; m++) if (bars[m].l < lo) lo = bars[m].l;
      move = price - lo;
    }
    out.push(Math.max(0, move) / a);
  }
  if (!out.length) return NaN;
  out.sort((x, y) => x - y);
  const mid = out.length >> 1;
  return out.length % 2 ? out[mid] : (out[mid - 1] + out[mid]) / 2;
}

function score(touches, spanBars, widthAtr, lookback, dist, allow, reaction, p) {
  const touchPts = Math.min(28, (touches - 2) * 10 + 9);
  const spanPts = Math.min(15, 15 * (spanBars / Math.max(lookback, 1)));
  const reactPts = (p.reactionFullAtr > 0 && Number.isFinite(reaction))
    ? 17 * Math.min(1, Math.max(0, reaction / p.reactionFullAtr)) : 0;
  const tightPts = p.maxWidthAtr <= 0 ? 0
    : 22 * Math.max(0, 1 - (widthAtr / p.maxWidthAtr));
  // scored against the SAME allowance the filter uses, so a zone that barely
  // survives the cut also scores near zero for closeness
  const proxPts = (allow <= 0 || !Number.isFinite(dist)) ? 0
    : 18 * Math.max(0, 1 - (dist / allow));
  return round2(Math.max(0, Math.min(100,
    touchPts + spanPts + tightPts + reactPts + proxPts)));
}

/** Agglomerate a price-sorted list, breaking wherever the gap exceeds `tol`. */
function cluster(levels, tol) {
  if (!levels.length) return [];
  const sorted = [...levels].sort((a, b) => a.price - b.price);
  const groups = [[sorted[0]]];
  for (const lv of sorted.slice(1)) {
    const g = groups[groups.length - 1];
    if (lv.price - g[g.length - 1].price <= tol) g.push(lv);
    else groups.push([lv]);
  }
  return groups;
}

/**
 * Zones visible at bar `i`, strongest first. Only pivots CONFIRMED by bar `i`
 * are used, so a swing is never counted before it became visible.
 */
export function detect(bars, i, timeframe, atrArr, params = {}) {
  const p = { ...DEFAULT_ZONE_PARAMS, ...params };
  const a = (i < atrArr.length && Number.isFinite(atrArr[i])) ? atrArr[i] : 0;
  if (a <= 0) return [];

  const lastClose = bars[i].c;
  const i0 = Math.max(0, i - p.lookback);
  /* HOW MANY BARS THIS ACTUALLY LOOKED AT, which is not always `lookback`.
     `i0` clamps at 0, so a window longer than the history silently becomes the
     history: on 4h the chart loads 1200 bars (js/util.js BAR_COUNT) and the
     1500-bar tier had been reporting 1500 of them. The replay clips it a second
     way -- its slice ends at the cursor, so 150 bars into a walk every tier is
     150 bars deep whatever it asked for. One field covers both, and it stays
     right if BAR_COUNT changes or history is extended on demand.
     `i - i0`, not `i - i0 + 1`: unclipped it then equals `lookback` exactly,
     so the number a reader sees matches the number the tier is named for. */
  const windowBars = i - i0;
  let winHi = -Infinity, winLo = Infinity;
  for (let k = i0; k <= i; k++) {
    if (bars[k].h > winHi) winHi = bars[k].h;
    if (bars[k].l < winLo) winLo = bars[k].l;
  }
  const allow = Math.max(p.maxDistanceAtr * a, p.maxDistanceRange * (winHi - winLo));
  /* AN INJECTION SEAM, and it changes nothing unless a caller uses it.
     `pivotsFn` defaults to exactly the call that was here, so the shipped
     behaviour and the parity tests are untouched. It exists because the pivot
     definition is the one input to this detector that has never been measured
     -- tools/zone_defn_audit.mjs feeds it a ZigZag instead -- and replicating
     the clustering inside a measurement tool would test a copy of this file
     rather than this file. */
  const { highs, lows } = (p.pivotsFn || findPivots)(bars, p.strengthPivots);

  /* findPivots does not carry confirmedI — it is shared with the batch scorer
     and must stay byte-identical for the other parity tests — so the confirming
     bar is derived here: it is i + strength by definition. */
  const conf = (q) => q.i + p.strengthPivots <= i && q.i >= i0;
  const levels = [
    ...highs.filter(conf).map((q) => ({ price: q.price, i: q.i, kind: 'high' })),
    ...lows.filter(conf).map((q) => ({ price: q.price, i: q.i, kind: 'low' })),
  ];
  if (levels.length < p.minTouches) return [];

  const times = bars.map((b) => b.t);
  const out = [];
  let seq = 0;

  for (const g of cluster(levels, p.clusterAtr * a)) {
    const gs = [...g].sort((x, y) => x.i - y.i);
    const kept = [];
    for (const lv of gs) {
      if (!kept.length || lv.i - kept[kept.length - 1].i >= p.minSeparation) kept.push(lv);
    }
    if (kept.length < p.minTouches) continue;
    const prices = kept.map((x) => x.price);
    const lo = Math.min(...prices), hi = Math.max(...prices);
    const widthAtr = (hi - lo) / a;
    if (widthAtr > p.maxWidthAtr) continue;
    const firstI = kept[0].i, lastI = kept[kept.length - 1].i;
    seq++;
    const z = new Zone({
      id: `${timeframe}-Z-${seq}`, timeframe, low: lo, high: hi,
      touches: kept.length, firstI, lastI,
      firstT: times[firstI], lastT: times[lastI],
      widthAtr,
      levels: prices.map((x) => Math.round(x * 1e10) / 1e10),
      fromHighs: kept.filter((x) => x.kind === 'high').length,
      fromLows: kept.filter((x) => x.kind === 'low').length,
      windowBars,
    });
    const dist = Math.abs(lastClose - z.mid);
    if (Number.isFinite(dist) && allow > 0 && dist > allow) continue;
    const reaction = reactionAtr(kept, bars, atrArr, i, p);
    z.reactionAtr = reaction;
    z.atr = a;
    z.strength = score(z.touches, lastI - firstI, widthAtr, p.lookback,
                       dist, allow, reaction, p);
    if (z.strength < p.minStrength) continue;
    out.push(z);
  }

  out.sort((x, y) => y.strength - x.strength);
  return out.slice(0, p.maxZones);
}

/** Zones for a chart's own bars — the entry point main.js uses. */
export function liveZones(bars, timeframe, params = {}) {
  if (!bars || bars.length < 60) return [];
  return detect(bars, bars.length - 1, timeframe, atrSeries(bars, 14), params);
}

/* ---------------------------------------------------------------------------
 * THE TIER LADDER -- one detector, three windows.
 *
 * WHY. `lookback` was 500 on every timeframe and had never been measured.
 * tools/zone_lookback_sweep.mjs swept 100..2000 across six cells and found two
 * things. The window is SCALE-FREE like everything else here -- 5m and 4h,
 * gold and cable all produce ~10 bands at 500, 0.30 ATR wide, within a point
 * of each other -- so the per-timeframe lookback table this was meant to
 * produce is unnecessary. And the shipped cap of six BINDS ON 72-80% OF BARS
 * at lookback 500. Three quarters of the time the chart was showing an
 * arbitrary six out of ten, ordered by a score already retired as a forecast.
 *
 * A ladder answers that better than a bigger cap. Six bands that mean three
 * different distances is a chart; ten bands that all mean the same thing is a
 * mesh. The tiers are the sweep's own three regimes:
 *
 *     near   100    the only window where the cap does not bind (0.6-1.9%);
 *                   two touches, median half an ATR from price -- what is
 *                   underfoot right now
 *     mid    500    today's shipped behaviour, kept as the middle rung so the
 *                   ladder is a superset of what the chart already drew
 *     far   1500    structural; a band still standing after 1500 bars, which
 *                   on 15m is most of a month
 *
 * A TIER IS CLAMPED BY THE HISTORY IT HAS, and says so. `lookback` is what the
 * tier ASKS for; `windowBars` on each zone is what it GOT. They differ on 4h,
 * 1d and 1w -- those load 1200 / 1000 / 800 bars, so the far tier is the whole
 * chart there -- and they differ everywhere early in a replay, whose slice ends
 * at the cursor. Below roughly 300 bars the far tier finds nothing the mid tier
 * has not already claimed and the ladder degrades to two rungs, which is the
 * dedup working rather than failing: fewer bands, not the same band three
 * times. The tooltip quotes `windowBars`, never the request.
 *
 * DEDUP IS BY SHARED PIVOT, and the NARROWEST window wins -- the tier is the
 * SHORTEST memory in which the level is visible. Matching on price instead
 * would split a band that gains a touch and moves its mid; the sweep made that
 * mistake first and invented churn out of it.
 *
 * THE OTHER DIRECTION WAS TRIED FIRST AND IS WRONG. Letting the widest window
 * claim a shared band starves `near` to nothing: the 500-bar window CONTAINS
 * the last 100 bars, so every recent cluster is also in the mid set and the
 * near tier never fires. Measured on the replay -- 4h, 1200 bars -- it
 * returned two `far` and two `mid` and not a single `near`.
 *
 * Narrowest-wins also says the more useful thing. A level a 100-bar memory can
 * already see is one price is working on NOW; a level that needs 1500 bars to
 * appear is old structure well away from price. That is the ladder as a reader
 * wants it -- immediate, medium, structural, spread across the screen by
 * distance -- rather than a re-labelling of the same near cluster.
 *
 * NOT MULTI-TIMEFRAME, still. A 4h band and a 15m band at one price are the
 * same band, so projecting one onto the other double-counts a level. A long
 * window on the chart's OWN bars is the same information without that.
 *
 * NO CLAIM ABOUT DIRECTION IS MADE OR IMPLIED. `far` is not stronger than
 * `near`; across 185,227 approaches nothing about a zone predicts which side
 * price leaves it, and a longer-lived band is no exception. The tier says how
 * far back you have to look to see the level -- a fact about the window, not
 * a forecast.
 */
export const ZONE_TIERS = [
  { tier: 'near', lookback: 100, keep: 2 },
  { tier: 'mid', lookback: 500, keep: 2 },
  { tier: 'far', lookback: 1500, keep: 2 },
];

/** Do two bands come from the same pivot cluster? */
function sharesPivot(a, b) {
  for (const x of a.levels) if (b.levels.includes(x)) return true;
  return false;
}

/**
 * Zones at bar `i` across all three windows, nearest tier first.
 *
 * Returns the same Zone objects `detect` does, each carrying `tier` and
 * `tierRank` (0 near .. 2 far). Callers that do not know about tiers see an
 * ordinary zone array.
 */
export function tieredDetect(bars, i, timeframe, atrArr, params = {}) {
  const out = [];
  // narrowest first, so the shortest window that saw a band is the one that keeps it
  for (let t = 0; t < ZONE_TIERS.length; t++) {
    const { tier, lookback, keep } = ZONE_TIERS[t];
    let zs;
    try {
      zs = detect(bars, i, timeframe, atrArr, { ...params, lookback, maxZones: 999 });
    } catch { continue; }
    let taken = 0;
    for (const z of zs) {
      if (taken >= keep) break;
      if (out.some((o) => sharesPivot(o, z))) continue;
      z.tier = tier;
      z.tierRank = t;
      z.tierLookback = lookback;
      out.push(z);
      taken++;
    }
  }
  return out;
}

/** Tiered zones for a chart's own bars -- the ladder's entry point. */
export function tieredZones(bars, timeframe, params = {}) {
  if (!bars || bars.length < 60) return [];
  return tieredDetect(bars, bars.length - 1, timeframe, atrSeries(bars, 14), params);
}

File: sim/tl/zones.py
"""
zones.py — horizontal support/resistance BANDS, not lines.

A line is a claim about one price. A zone is a claim about a region, and it is
the more honest object: price does not turn at 1.16847, it turns somewhere in a
band a few tenths of an ATR wide, and every chart annotated by a human draws it
that way -- "Supply", "Demand", "Seller Zone", "Buyer Zone".

HOW A ZONE IS BUILT

  1. Every confirmed swing pivot inside the lookback is a candidate level.
  2. Levels within `cluster_atr` of each other are one zone. Agglomerative on a
     sorted list, which is the right shape here: levels are one-dimensional, and
     merging nearest-neighbours in price order gives the same answer as anything
     more elaborate at a fraction of the cost.
  3. A zone must have `min_touches` distinct pivots from at least
     `min_separation` bars apart. Three pivots inside one week's consolidation
     is one event seen three times, not three independent tests.

WHY PIVOTS AND NOT LINE VALUES. A trendline's value moves with time, so
clustering line values would produce a band that means something only at the
instant it was computed. Pivots are fixed prices. A zone is horizontal by
definition -- if the structure slopes, that is a channel, and channels.py
already has it.

ROLE IS ASSIGNED FROM PRICE, NOT FROM THE PIVOT TYPE. A zone built from swing
highs is resistance while price is below it and support once price is above it:
that flip is the single most-taught idea in support/resistance, and hard-coding
the role from the pivots would throw it away. `role_at(price)` decides.

STRENGTH combines how many times the zone was tested, how far apart those tests
were in time, and how TIGHT the cluster is -- a band 0.15 ATR wide that price
turned at four times is a very different object from one 0.9 ATR wide that also
has four touches, and averaging them into "4 touches" loses the distinction.

CAUSALITY: pure function of pivots confirmed at or before bar `i`, and bars up
to `i`. Nothing later is read because nothing later is passed.
"""

from dataclasses import dataclass
from typing import List, Optional

import numpy as np

from .pivots import find_pivots, pivots_confirmed_by

SUPPORT = 'support'
RESISTANCE = 'resistance'


@dataclass
class Zone:
    """A horizontal band price has repeatedly reacted to."""

    id: str
    timeframe: str
    low: float
    high: float
    touches: int
    first_i: int
    last_i: int
    first_t: int
    last_t: int
    width_atr: float = 0.0
    strength: float = 0.0
    #: median excursion away from this zone's pivots, in ATR
    reaction_atr: float = float('nan')
    atr: float = float('nan')     # ATR at detection; lets a UI talk in points
    # Pivot prices that formed it, so a chart can mark the evidence.
    levels: tuple = ()
    from_highs: int = 0
    from_lows: int = 0

    @property
    def mid(self) -> float:
        return 0.5 * (self.low + self.high)

    def contains(self, price) -> bool:
        return self.low <= price <= self.high

    def role_at(self, price) -> str:
        """
        Resistance while price is below the band, support once it is above.
        Inside the band the nearer edge decides, because that is the edge the
        next move has to clear.
        """
        if price < self.low:
            return RESISTANCE
        if price > self.high:
            return SUPPORT
        return SUPPORT if (price - self.low) > (self.high - price) else RESISTANCE

    def distance_atr(self, price, atr) -> float:
        if not atr:
            return float('nan')
        if self.contains(price):
            return 0.0
        d = (self.low - price) if price < self.low else (price - self.high)
        return d / atr

    def to_row(self) -> dict:
        return {
            'id': self.id, 'timeframe': self.timeframe,
            'low': self.low, 'high': self.high, 'mid': self.mid,
            'touches': self.touches, 'from_highs': self.from_highs,
            'from_lows': self.from_lows,
            'first_t': self.first_t, 'last_t': self.last_t,
            'width_atr': round(self.width_atr, 3),
            'strength': self.strength,
            'reaction_atr': self.reaction_atr,
            'atr': self.atr,
        }


@dataclass
class ZoneParams:
    strength_pivots: int = 3      # fractal size for the pivots that feed it
    close_confirm: bool = False   # see sim/tl/pivots.py find_pivots
    lookback: int = 500           # bars considered
    cluster_atr: float = 0.35     # levels closer than this are one zone
    # 2 rather than 3. A clean double top IS a level a human draws, and the
    # scorer already ranks it accordingly -- touch_pts gives 2 touches 11 points
    # against 35 for four -- so a weak zone loses on merit instead of being
    # excluded before it can compete. Hard-gating at 3 removed 23 of 33 clusters
    # on gold H1 before anything had a chance to be judged.
    min_touches: int = 2          # distinct reactions required
    min_separation: int = 8       # bars apart before two touches count as two
    max_width_atr: float = 1.2    # wider than this is not a level, it is a range
    max_zones: int = 6            # best first
    min_strength: float = 25.0
    # How far price travelled AWAY from a pivot before turning back, in ATR,
    # measured over this many bars. A level that produced 3 ATR bounces is not
    # the same object as one price grazed and drifted from, and a raw touch
    # count cannot tell them apart.
    reaction_bars: int = 20
    reaction_full_atr: float = 2.0    # the excursion that earns full marks
    # HOW FAR IS TOO FAR. A zone price cannot reach is true and untradeable,
    # but "12 ATR" turned out to mean wildly different things per instrument:
    # measured over the same 500-bar window it covered 19% of the actual price
    # range on USDJPY 1h and 54% on XAUUSD 4h. A trader does not think in ATR
    # here -- they draw the levels that are ON THE CHART.
    #
    # So the allowance is the LARGER of an ATR budget and a fraction of the
    # range price has actually covered in the lookback. The range term is what
    # normally binds and self-scales with the instrument; the ATR term is a
    # floor that keeps the allowance sane through an unusually quiet window,
    # where the recent range can collapse to almost nothing.
    max_distance_atr: float = 12.0
    max_distance_range: float = 0.75


def _score(touches, span_bars, width_atr, lookback, dist, allow, reaction,
           p: ZoneParams) -> float:
    """
    0-100 over five terms.

    TIGHTNESS is weighted heavily because it separates a zone from a vague area:
    a band price turned at four times within 0.15 ATR is a level, the same four
    touches spread over a full ATR is a neighbourhood.

    REACTION is what a raw touch count cannot express. Three pivots that each
    produced a 3 ATR bounce and three that barely turned score the same on
    `touches` and are not the same object -- the first is a level being defended,
    the second is price passing through a price. Weight had to come from
    somewhere: touches gives up 7 points and span 5, because both were already
    proxies for the same underlying thing this measures directly.

    Distance discounts the whole thing last: a level you cannot reach today is
    not a level you can trade today.
    """
    touch_pts = min(28.0, (touches - 2) * 10.0 + 9.0)
    span_pts = min(15.0, 15.0 * (span_bars / max(lookback, 1)))
    react_pts = 0.0
    if p.reaction_full_atr > 0 and np.isfinite(reaction):
        react_pts = 17.0 * min(1.0, max(0.0, reaction / p.reaction_full_atr))
    if p.max_width_atr <= 0:
        tight_pts = 0.0
    else:
        tight_pts = 22.0 * max(0.0, 1.0 - (width_atr / p.max_width_atr))
    # Proximity is scored against the SAME allowance the filter uses, so a zone
    # that just survives the cut also scores near zero for closeness. Scoring it
    # against a different yardstick would let a zone be "far" for one purpose
    # and "near" for the other.
    if allow <= 0 or not np.isfinite(dist):
        prox_pts = 0.0
    else:
        prox_pts = 18.0 * max(0.0, 1.0 - (dist / allow))
    return round(max(0.0, min(100.0, touch_pts + span_pts + tight_pts
                              + react_pts + prox_pts)), 2)


def _reaction_atr(kept, high, low, atr, i, p: ZoneParams) -> float:
    """
    Median excursion away from the zone's pivots, in ATR.

    For a swing LOW the reaction is how far price rose afterwards; for a swing
    HIGH how far it fell. Median rather than mean, so one violent bounce cannot
    carry a level that otherwise did nothing.

    CAUSAL. Every bar read lies between the pivot and `i`, both already in the
    past at the moment the zone is computed. The window is also clipped at `i`,
    so a recent pivot is scored on the bars that actually exist rather than
    being credited with an excursion that has not happened yet.
    """
    out = []
    for price, k, kind in kept:
        a = atr[k] if k < len(atr) and np.isfinite(atr[k]) and atr[k] > 0 else np.nan
        if not np.isfinite(a):
            continue
        end = min(i, k + p.reaction_bars)
        if end <= k:
            continue
        if kind == 'low':
            move = float(np.max(high[k + 1:end + 1])) - price
        else:
            move = price - float(np.min(low[k + 1:end + 1]))
        out.append(max(0.0, move) / a)
    return float(np.median(out)) if out else float('nan')


def _cluster(levels, tol):
    """
    Agglomerate a price-sorted list: walk it once, starting a new group whenever
    the gap to the previous level exceeds `tol`. `levels` is [(price, i, kind)].
    """
    if not levels:
        return []
    levels = sorted(levels, key=lambda x: x[0])
    groups = [[levels[0]]]
    for lv in levels[1:]:
        if lv[0] - groups[-1][-1][0] <= tol:
            groups[-1].append(lv)
        else:
            groups.append([lv])
    return groups


def detect(bars, i, timeframe, atr, params: ZoneParams = None, times=None):
    """
    Zones visible at bar `i`, strongest first.

    Only pivots CONFIRMED by bar `i` are used, so a swing is never counted
    before the bar it actually became visible.
    """
    p = params or ZoneParams()
    a = float(atr[i]) if i < len(atr) and np.isfinite(atr[i]) else 0.0
    if a <= 0:
        return []

    high = np.asarray(bars['high'], dtype=float)
    low = np.asarray(bars['low'], dtype=float)
    close = np.asarray(bars['close'], dtype=float)
    last_close = float(close[i])
    if times is None:
        times = np.asarray(bars.index.astype('int64') // 1_000_000)

    i0 = max(0, i - p.lookback)
    # The distance allowance, computed once: the larger of the ATR budget and a
    # fraction of the range price actually covered in the lookback.
    win_hi = float(np.max(high[i0:i + 1]))
    win_lo = float(np.min(low[i0:i + 1]))
    allow = max(p.max_distance_atr * a, p.max_distance_range * (win_hi - win_lo))
    piv_hi, piv_lo = find_pivots(high, low, p.strength_pivots, close=close,
                                 close_confirm=getattr(p, 'close_confirm', False))
    piv_hi = [q for q in pivots_confirmed_by(piv_hi, i) if q['i'] >= i0]
    piv_lo = [q for q in pivots_confirmed_by(piv_lo, i) if q['i'] >= i0]

    levels = [(q['price'], q['i'], 'high') for q in piv_hi]
    levels += [(q['price'], q['i'], 'low') for q in piv_lo]
    if len(levels) < p.min_touches:
        return []

    out = []
    seq = 0
    for g in _cluster(levels, p.cluster_atr * a):
        # distinct touches: collapse pivots that sit within min_separation bars
        g_sorted = sorted(g, key=lambda x: x[1])
        kept = []
        for lv in g_sorted:
            if not kept or lv[1] - kept[-1][1] >= p.min_separation:
                kept.append(lv)
        if len(kept) < p.min_touches:
            continue
        prices = [x[0] for x in kept]
        lo, hi = min(prices), max(prices)
        width_atr = (hi - lo) / a
        if width_atr > p.max_width_atr:
            continue
        first_i, last_i = kept[0][1], kept[-1][1]
        seq += 1
        z = Zone(
            id='%s-Z-%d' % (timeframe, seq),
            timeframe=timeframe, low=lo, high=hi,
            touches=len(kept), first_i=first_i, last_i=last_i,
            first_t=int(times[first_i]), last_t=int(times[last_i]),
            width_atr=width_atr,
            levels=tuple(round(x, 10) for x in prices),
            from_highs=sum(1 for x in kept if x[2] == 'high'),
            from_lows=sum(1 for x in kept if x[2] == 'low'),
        )
        dist = abs(last_close - z.mid)
        if np.isfinite(dist) and allow > 0 and dist > allow:
            continue
        reaction = _reaction_atr(kept, high, low, atr, i, p)
        z.reaction_atr = reaction
        z.atr = a
        z.strength = _score(z.touches, last_i - first_i, width_atr, p.lookback,
                            dist, allow, reaction, p)
        if z.strength < p.min_strength:
            continue
        out.append(z)

    out.sort(key=lambda z: -z.strength)
    return out[:p.max_zones]
